src/part_file.cpp

```diff
@@ -94,13 +94,13 @@ namespace libtorrent
 
 		error_code ec;
 		std::string fn = combine_path(m_path, m_name);
-		m_file.open(fn, file::read_only, ec);
+		boost::shared_ptr<file> const f = boost::make_shared<file>(fn, file::read_only, boost::ref(ec));
 		if (!ec)
 		{
 			// parse header
 			boost::scoped_array<boost::uint32_t> header(new boost::uint32_t[m_header_size / 4]);
 			file::iovec_t b = {header.get(), size_t(m_header_size) };
-			int n = m_file.readv(0, &b, 1, ec);
+			int const n = f->readv(0, &b, 1, ec);
 			if (ec) return;
 
 			// we don't have a full header. consider the file empty
```

Comment:
I can't see previous comments on this.
I'll duplicate the main point, I think it is fine to release m_file, because subsequent reads from an empty/invalid file will fail. Removing such file from disk when possible is also fine.

Reply:
No problem.

Releasing the file is not that important on most modern operating systems. At the same time I suggest to avoid wasting of system resources when possible.

Since `m_piece_map` is always verified before reading, `open_file` won't be called for empty/invalid file.

Reply:
in fact, since the part file is only expected to be used rarely, perhaps it doesn't make sense to keep an open file at all, but to just open it on demand. it would make it even "more" thread safe. maybe something for the master branch

Reply:
oh, but this comment still applies to the constructor. your point has been addressed. Note how the file is opened and owned by a local variable. Only after everything succeeded will the `m_file` be assigned

Reply:
Do you think invalid/empty file could be removed in constructor?

Reply:
you mean if there is a file on disk already, but it's empty?
as long as any potential races could be worked out, I suppose so. There shouldn't be any race in the constructor I wouldn't think. I think that's for another PR though

Reply:
Exactly.

I agree, this question if out of current scope.